docs/products/kafka/concepts/log-compaction.rst

```diff
@@ -1,7 +1,169 @@
 
+
+How topic log compaction works
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+ Apache Kafka topics represent a continuous stream of messages that typically get discarded after log reaches a certain period of time or size. However for certain use cases we just need the most recent changes.
+
+For example, if we have a topic that contains user's home address and every time there is an update, it gets sent to the topic using ``user_id`` as primary key and home address as the value:
+
+::
+
+   1001 -> "4 Privet Dr"
+   1002 -> "221B Baker Street"
+   ...
+   1001 -> "21 Jump St"
+   ...
+   1001 -> "Paper St"
+
+
+We have three different options on how long to retain the messages:
+
+* infinite log retention: all changes to user's address would be maintained in the logs. This would lead to the log growing in size without a bound.
+* simple log retention: older log records would be deleted after log reaches certain age or size.
+* log compaction: only latest version of the key's value are kept.
+
+With log compaction Apache Kafka would remove any older records for which there is a newer version available in the partition log. This retention policy can be set per-topic, so a single cluster can have some topics where retention is enforced by size or time and other topics where retention is enforced by compaction.
+
+Log compaction basics
+~~~~~~~~~~~~~~~~~~~~~
+
+To understand better how log compaction work we will look at a partition log of a compacted topic before and after compaction has been applied.
+
+Before compaction
+*****************
+
+.. list-table::
+  :header-rows: 1
+  :stub-columns: 1
+  :align: left
+
+  * - Offset
+    - Key
+    - Value
+  * - 1
+    - K1
+    - 13
+  * - 2
+    - K2
+    - 11
+  * - 3
+    - K3
+    - 22
+  * - 4
+    - K3
+    - 7
+  * - 5
+    - K1
+    - 8
+  * - 6
+    - K4
+    - 14
+
+You can notice that there are two records with duplicate keys **K1**  and **K3.** After we apply log compaction, we only keep records with the latest offset (newest values) and the older ones get discarded.
+
+After compaction
+*****************
+
+.. list-table::
+  :header-rows: 1
+  :stub-columns: 1
+  :align: left
+
+  * - Offset
+    - Key
+    - Value
+  * - 2
+    - K2
+    - 11
+  * - 4
+    - K3
+    - 7
+  * - 5
+    - K1
+    - 8
+  * - 6
+    - K4
+    - 14
+
+
+When a log is compacted it consists of head and tail, where head is the traditional  Apache Kafka log and new records get appended to the end of it.  Apache Kafka ensures that the records in the tail consist only of unique keys because only the tail section is scanned during compaction process while head section may contain duplicate keys.
+
+.. note:: Log compaction occurs inside a partition and if two records with the same key land in different partitions, they will not be compacted together.
```

Comment:
Should we state that "this usually doesn't happen since the record key is used to select the partition, but for custom message routing this might be an issue"

Reply:
Added